distributed/shuffle/shuffle_extension.py

```diff
@@ -329,3 +429,151 @@ def _get_shuffle(self, shuffle_id: ShuffleId) -> Shuffle:
             raise ValueError(
                 f"Shuffle {shuffle_id!r} is not registered on worker {self.worker.address}"
             ) from None
+
+    async def close(self):
+        self.executor.shutdown()
+        MultiFile.queue = None
+
+
+class ShuffleSchedulerExtension:
+    """
+    Shuffle extension for the scheduler
+
+    Today this mostly just collects heartbeat messages for the dashboard,
+    but in the future it may be responsible for more
+
+    See Also
+    --------
+    ShuffleWorkerExtension
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.shuffles = defaultdict(lambda: defaultdict(dict))
+
+    def heartbeat(self, ws, data):
+        for shuffle_id, d in data.items():
+            self.shuffles[shuffle_id][ws.address].update(d)
+
+
+def split_by_worker(
```

Comment:
It would be nice to move all of these functions to a separate file.  mypy was sad when I did this though and I couldn't figure out why, which is why they're still here.

Reply:
@crusaderky if you have 30m on Monday afternoon your time for a quick screenshare I could use your expertise here.  I suspect that with you helping I could learn enough to work around this issue quickly.

Reply:
Sure, feel free to drop me an invite